### DISCUSSION TITLE
How to print console.log statements in tests to test output in terminal

### DISCUSSION THREAD

**Discussion by jhhom** [vitest-dev/vscode#117] [Q&A]
Hi, thanks for the great extension.

One problem I'm facing is if I have a `console.log` statements in my tests, it doesn't get outputted to the terminal, compared to the command `npx vitest` which does output `console.log` statements in terminal

My setting for `vitest.commandLine` is `npx vitest --watch`.

<img width="560" alt="image" src="https://user-images.githubusercontent.com/24519631/211549326-1abf2d2f-4852-4eee-b9f1-92bb8ecdb0b0.png">

Here is the test output I'm getting when running the extension. May I know how to get the `log` statements printed to test output?

<img width="424" alt="image" src="https://user-images.githubusercontent.com/24519631/211549558-3c9a0c1a-6351-4183-9d90-3473f402d03a.png">

This is what I get from running `npx vitest` instead, which does print the `console.log` statements
<img width="688" alt="image" src="https://user-images.githubusercontent.com/24519631/211550458-3098fd4a-3d0b-4867-ba9d-843e0ae0a968.png">

What I found is if I changed the `viteset.commandLine` setting to `npx vitest --` instead, it will show the console log output, however the extension will hang on the test and never finished as the command "Wait for file changes" forever.

<img width="480" alt="image" src="https://user-images.githubusercontent.com/24519631/211553090-e8b31707-8b17-447e-9dd6-7b73aa97ee5a.png">

<img width="264" alt="image" src="https://user-images.githubusercontent.com/24519631/211552949-437bb5af-4070-46dc-aeb2-e6305e457882.png">

So is it possible to get the `console` log outputs with the `npx vitest --watch` command? Or how to make the extension works properly while still getting the `console` log statements printed on the terminal?

**Comment by jossephus** [comment]
It's been sometime but for anyone facing the same problem, you can set --silent=false when running vitest to get console outputs.

**Reply by castarco** [reply]
This does not work at all.

**Reply by neilhuyton** [reply]
Worked for me.

**Reply by davem-mf** [reply]
If you're running through `npm t`, make sure you put in the `--` to kick the flag to vitest:

Good:
```
npm t -- --silent=false
```

**Comment by jhhom** [comment]
<img width="527" alt="image" src="https://user-images.githubusercontent.com/24519631/235908523-4e61691d-7f97-48f0-bdf4-f069cb1fd970.png">

Is this setting correct? I'm still not getting the console outputs on my terminal from running the test using the extension

**Comment by snuup** [comment]
No results here as well. Can someone please help us?

**Comment by DanTup** [comment]
I just hit this too.. I found `--disable-console-intercept` documented on https://vitest.dev/guide/cli which seems to be working for me.

**Reply by tankorsmash** [reply]
Worked for me, thank you!

```
pnpx vitest --inspect-brk --no-file-parallelism --disable-console-intercept 
```

**Reply by tomardern** [reply]
Worked for me too! thanks :)

**Comment by RangerMauve** [comment]
I found that with the default reporter it swallows any output anyways. Using `--reporter=json` along with `--disable-console-intercept` worked.

**Comment by kevireilly** [comment]
Appears as though you can also use `--reporter=basic --disable-console-intercept` to get the same `default` reporter output while including console output.

**Comment by iamsalmakolkari** [comment]
Vitest provides a configuration method called `onConsoleLog` which is Custom handler for console.log in tests. It is basically used if you want to supress logs based on the message text. So, if you want to see console logs add below code in vitest.config.ts.
```typescript
/// <reference types="vitest" />
import { defineConfig, mergeConfig } from 'vitest/config';
import viteConfig from './vite.config';

export default mergeConfig(
  viteConfig({ command: 'build', mode: 'TEST' }),
  defineConfig({
    test: {
      coverage: {
        exclude: ['src/api'],
        include: ['src'],
      },
      environment: 'jsdom',
      globals: true,
      setupFiles: './setupTests.ts',
      onConsoleLog(log: string, type: 'stdout' | 'stderr'): false | void {
        console.log('log in test: ', log);
        if (log === 'message from third party library' && type === 'stdout') {
          return false;
        }
      },
    },
  }),
);

```

Quick note: If you have spied on console.log for testing any scenarios in your test case, then the spy will swallow the logs and no log in your test case will be printed to terminal.

**Comment by Caalb** [comment]
In my `vitest.config.js` I had a config called `pool: ‘vmThreads’`, removing it I was able to display the consoles correctly

**Comment by christophemarois** [comment]
Trick for certain situations: if you're just trying to get something printed to stdout, `process.stdout.write()` will work.

**Comment by pgraham-calix** [comment]
Appreciate all of the work arounds, but I think it would be an important (maybe necessary) feature of a test runner to be able to just log output without having to mess with config.

**Reply by pgraham-calix** [reply]
Actually, looks like it does work out of the box, I just missed what I was longing amongst the rest of the output.

**Reply by sheremet-va** [reply]
The extension does print the output. In fact, it only prints the user's output, ignoring the Vitest CLI reporting

**Comment by b4lch** [comment]
For anybody stuck using vitest v1.x, you can use `--silent=true --reporter=basic` as the default reporter collapses the output of your previous tests.

**Comment by albertreed** [comment]
There are several answers here - none of which seem to be working for me. Is there a definitive way to enable console.log() printing? As a previous poster mentioned, this feels like an essential thing to allow.

**Reply by nidomiro** [reply]
for me it did also not work until I changed `npm test --disable-console-intercept` to `npm test -- --disable-console-intercept`

**Comment by CyberT33N** [comment]
The above solutions does not work with console.info

**Comment by ddzy** [comment]
`console.log()` is not working
Try `window.console.log()` working for me

**Comment by nidomiro** [comment]
for me it did also not work until I changed `npm test --disable-console-intercept` to `npm test -- --disable-console-intercept`

**Comment by paul-uz** [comment]
`console.log()` fails, despite all the above setting variations. `console.warn()` does work

**Comment by steve-scorn** [comment]
If you're wanting the logging to appear in the JetBrains tools such as WebStorm try a `vitest.config.ts` like this one:

```
export default defineConfig({
    test: {
        environment: 'node',
        globals: true,
        include: ['**/*.test.ts'],
        disableConsoleIntercept: true
    },
    resolve: {
        alias: {
            '@': path.resolve(__dirname, './src')
        }
    }
});
```

It's the `disableConsoleIntercept: true` that did the trick for me.

**Comment by abenhamdine** [comment]
none of the differents solutions described here has worked for me (I'm on vitest 4, if matters)

**Reply by paramsiddharth** [reply]
I'm so tired and sick of this, I just ended up writing the logs to a file instead.

**Comment by CyberT33N** [comment]
Use loggers like pino or winston instead. Then it should work.

However, this should also work
https://vitest.dev/guide/cli.html#disableconsoleintercept

If not try other console like warn, debug or error instead of log

**Comment by gass-git** [comment]
The vitest option ` --disable-console-intercept` is now `--disableConsoleIntercept` as you can see in the [docs here](https://vitest.dev/guide/cli.html#disableconsoleintercept)

So, try (if you are using npm):

```
npx vitest --disableConsoleIntercept
```

If you are using pnpm:
```
pnpm vitest --disableConsoleIntercept
```

Or in your package.json:
```
"scripts": {
   "tests": "vitest --disableConsoleIntercept"
}
```
